distribution/xfer/download.go

```diff
@@ -183,15 +183,16 @@ func (ldm *LayerDownloadManager) Download(ctx context.Context, initialRootFS ima
 		progress.Update(progressOutput, descriptor.ID(), "Pulling fs layer")
 
 		var xferFunc DoFunc
-		cacheDir := ldm.cacheDirManager.GetDir(transferKey)
-		cacheDirs[transferKey] = cacheDir
 		if topDownload != nil {
-			xferFunc = ldm.makeDownloadFunc(descriptor, "", topDownload, cacheDir)
+			xferFunc = ldm.makeDownloadFunc(descriptor, "", topDownload, transferKey)
 			defer topDownload.Transfer.Release(watcher)
 		} else {
-			xferFunc = ldm.makeDownloadFunc(descriptor, rootFS.ChainID(), nil, cacheDir)
+			xferFunc = ldm.makeDownloadFunc(descriptor, rootFS.ChainID(), nil, transferKey)
 		}
 		topDownloadUncasted, watcher = ldm.tm.Transfer(transferKey, xferFunc, progressOutput)
+		if watcher.IsOriginal() {
+			cacheDirs[transferKey] = struct{}{}
+		}
```

Comment:
Not really a fan of this. Why can't the `ReleaseDir` calls be in the closure returned by `makeDownloadFunc`, after it registers the layer?